Jenkinsfile

```diff
@@ -0,0 +1,124 @@
+pipeline {
+    agent {
+        label 'linux && python'
+    }
+    stages {
+        stage("Version Check") {
+            steps {
+                checkVersion(discoverVersion())
+            }
+        }
+        stage('Build') {
+            steps {
+                echo 'Building..'
+                sh '''
+                    virtualenv ./sdk-deploy
+                    . ./sdk-deploy/bin/activate
+                    python setup.py sdist
+                    python setup.py bdist_wheel
+                '''
+            }
+        }
+        stage('Test') {
+            agent {
+                dockerfile {
+                    label 'linux'
+                }
+            }
+            steps {
+                echo 'Testing..'
+                sh '''
+                    virtualenv ./sdk-test
+                    . ./sdk-test/bin/activate
+                    tox
+                '''
+            }
+        }
+        stage('Deploy') {
+            steps {
+                echo 'Deploying....'
+                sh '''
+                    . ./sdk-deploy/bin/activate
+                    python -m pip install twine
+                    twine upload dist/* -u ${PYPI_USER_ID} -p ${PYPI_PASSWORD}
+                '''
+                tagRepo(discoverVersion())
+            }
+        }
+    }
+    post {
+        success {
+            notifyComplete('SUCCESS')
+        }
+        failure {
+            notifyComplete('FAILURE')
+        }
+    }
+}
+
+// Get version number from PKG-INFO file
+def discoverVersion() {
+    def pkgFile = readFile("src/rev_ai.egg-info/PKG-INFO")
+    versionLine = pkgFile.split('\n')[2] // get line with version number
+    assert versionLine =~ /Version(.*)/
+    return versionLine.substring(versionLine.lastIndexOf(':')+2, versionLine.length())
+}
+
+// Ensure that the version number has been incremented since last release
+def checkVersion(version) {
+    def tags = sh (script: "git ls-remote https://${env.GIT_CREDENTIALS}@${getRepoUrl()} --tags origin ${version}", returnStdout: true)
+    if (tags.trim() != "") {
+        error "${version} is already released, please increase the version number for build to publish"
+    }
+}
+
+// Tag repo with version number so we know this version has been released
+def tagRepo(tag) {
+    echo "tagging repo with ${tag}"
+
+    // deletes current tag locally, this will make sure if the tag is deleted from remote it would be set again
+    sh "git tag -d ${tag} || (exit 0)"
+    // see if it's still on remote - needed to avoid overwriting the tag
+    sh "git fetch https://${env.GIT_CREDENTIALS}@${getRepoUrl()} --tags"
```

Comment:
what's the purpose of this line?

Reply:
I'm actually not sure. The comment says it checks if the tag is still on remote but we don't do anything with taht information. These portions of the script are copied from revcommon